File: P3_Mojo/math/Nsieve/nsieve.py
import sys
import time


def nsieve(n):
    count = 0
    flags = [True] * n
    for i in range(2, n):
        if flags[i]:
            count += 1
            for j in range(i << 1, n, i):
                flags[j] = False
    print(f'Números primos hasta {n:8} {count:8}')


if __name__ == '__main__':
    # Obtener el valor límite desde los argumentos de línea de comandos
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 4

    start_time = time.time()
    # Calcular y mostrar los números primos para tres valores diferentes
    for i in range(0, 3):
        nsieve(10000 << (n-i))
    end_time = time.time()
    duration = end_time - start_time
    print(f"Tiempo de ejecución: {duration:.6f} segundos")

File: P3_Mojo/math/Nsieve/nsieve.cpp
#include <iomanip>
#include <iostream>
#include <sstream>
#include <vector>
#include <chrono>

void nsieve(std::size_t max) {
  // Crear vector para marcar los números no primos
  static std::vector<bool> no_primos;
  no_primos.assign(max, false);
  std::size_t cuenta = 0;

  // Calcular los números primos utilizando la criba de Eratóstenes
  for (std::size_t valor = 2; valor < max; ++valor) {
    if (!no_primos[valor]) {
      ++cuenta;

      // Marcar los múltiplos del número actual como no primos
      for (std::size_t multiple = valor * 2; multiple < max; multiple += valor) {
        no_primos[multiple] = true;
      }
    }
  }

  // Mostrar la cantidad de números primos encontrados
  std::cout << "Primos hasta " << std::setw(8) << max << ' ' << std::setw(8)
            << cuenta << '\n';
}

int main(int argc, char **argv) {
  if (argc != 2) {
    // Verificar la cantidad de argumentos
    std::cerr << "uso: " << argv[0] << " <n>\n";
    return 1;
  }

  // Convertir el argumento a un número entero
  unsigned int limite;
  {
    std::istringstream convertor(argv[1]);
    if (!(convertor >> limite) || !convertor.eof()) {
      std::cerr << "uso: " << argv[0] << " <n>\n";
      std::cerr << "\tn debe ser un número entero\n";
      return 1;
    }
  }

  // Obtener el tiempo de inicio
  auto inicio = std::chrono::high_resolution_clock::now();

  // Calcular y mostrar los números primos para tres valores diferentes
  for (std::size_t i = 0; i < 3; ++i) {
    nsieve(10000 << (limite - i));
  }

  // Obtener el tiempo de finalización y calcular la duración
  auto fin = std::chrono::high_resolution_clock::now();
  std::chrono::duration<double> duracion = fin - inicio;

  // Mostrar la duración de la ejecución
  std::cout << "Tiempo de ejecución: " << duracion.count() << " segundos\n";

  return 0;
}
